infrastructure/src/main/java/org/corfudb/infrastructure/CompactorLeaderServices.java

```diff
@@ -0,0 +1,485 @@
+     *      ... monitor checkpointing of the table by observing if the checkpointStream's tail moves forward
+     *      ... if it does not move forward for timeout ms, then mark it as failed
+     * Also, when checkpoint of a table is found to be failed, the cycle is immediately marks as failed.
+     * @param timeout in ms
+     */
+    public void validateLiveness(long timeout) {
+        List<TableName> tableNames;
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            tableNames = new ArrayList<>(txn.keySet(activeCheckpointsTable)
+                    .stream().collect(Collectors.toList()));
+            txn.commit();
+        } catch (Exception e) {
+            syslog.warn("Unable to acquire tableNames");
+            return;
+        }
+
+        long currentTime = System.currentTimeMillis();
+        if (tableNames.size() == 0) {
+            handleNoActiveCheckpointers(timeout, currentTime);
+        }
+        for (TableName table : tableNames) {
+            LivenessValidatorHelper.clear();
+            String streamName = TableRegistry.getFullyQualifiedTableName(table.getNamespace(), table.getTableName());
+            UUID streamId = CorfuRuntime.getCheckpointStreamIdFromName(streamName);
+            long currentStreamTail = corfuRuntime.getSequencerView()
+                    .getStreamAddressSpace(new StreamAddressRange(streamId, Address.MAX, Address.NON_ADDRESS)).getTail();
+
+            long syncHeartBeat = -1;
+            try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+                syncHeartBeat = txn.getRecord(activeCheckpointsTable, table).getPayload().getSyncHeartbeat();
+                txn.commit();
+            } catch (Exception e) {
+                syslog.warn("Unable to acquire table status");
+            }
+            if (readCache.containsKey(table)) {
+                //previousStatus.first is syncHeartbeat
+                //previousStatus.second is StreamTail
+                Tuple<Long, Long> previousStatus = readCache.get(table).first;
+                if (previousStatus.second < currentStreamTail) {
+                    readCache.put(table, Tuple.of(Tuple.of(previousStatus.first, currentStreamTail), currentTime));
+                } else if (previousStatus.first < syncHeartBeat) {
+                    readCache.put(table, Tuple.of(Tuple.of(syncHeartBeat, previousStatus.second), currentTime));
+                } else if (previousStatus.second == currentStreamTail || previousStatus.first == syncHeartBeat) {
+                    if (currentTime - readCache.get(table).second > timeout && !handleSlowCheckpointers(table)) {
+                        readCache.clear();
+                        return;
+                    }
+                }
+            } else {
+                readCache.put(table, Tuple.of(Tuple.of(syncHeartBeat, currentStreamTail), currentTime));
+            }
+        }
+    }
+
+    private int findCheckpointProgress() {
```

Comment:
Can we use executeQuery to simply this method? Something like 
`txn.executeQuery(checkpointingStatusTable, 
record -> ((CheckpointingStatus) record.getPayload()).getStatus() == StatusType.IDLE)`
Also can we change the method name to `getIdleCount()`?